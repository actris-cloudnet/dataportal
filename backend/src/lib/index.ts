export const stringify = (obj: any): string => JSON.stringify(obj, null, 2)
