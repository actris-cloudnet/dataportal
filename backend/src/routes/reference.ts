import { Connection, Repository } from "typeorm";
import { Request, RequestHandler, Response, NextFunction } from "express";

import { RegularFile, ModelFile } from "../entity/File";
import { Collection } from "../entity/Collection";
import axios from "axios";

const LABELLING_URL = "https://actris-nf-labelling.out.ocp.fmi.fi/api/facilities";

interface Name {
  first_name: string;
  last_name: string;
}

interface Citation {
  citekey: string;
  author: Name[];
  publisher: string[];
  title: string;
  year: string;
  month: string;
  day: string;
  url: string;
  urldate: string;
  note: string;
}

interface Reference {
  citation: Citation;
  acknowledgements: string;
  dataAvailability: string;
}

export class ReferenceRoutes {
  private fileRepository: Repository<RegularFile>;
  private modelRepository: Repository<ModelFile>;
  private collectionRepository: Repository<Collection>;

  constructor(conn: Connection) {
    this.fileRepository = conn.getRepository<RegularFile>("regular_file");
    this.modelRepository = conn.getRepository<ModelFile>("model_file");
    this.collectionRepository = conn.getRepository<Collection>("collection");
  }

  getReference: RequestHandler = async (req: Request, res: Response, next) => {
    let data: RegularFile | ModelFile | undefined = await this.fileRepository.findOne(
      { uuid: req.params.uuid },
      { relations: ["site", "site.citations", "product"] }
    );
    if (!data) {
      data = await this.modelRepository.findOne(
        { uuid: req.params.uuid },
        { relations: ["site", "site.citations", "product", "model", "model.citations"] }
      );
    }
    if (data === undefined) {
      return next({ status: 404, errors: "UUID not found" });
    }
    const pi = await getPrincipalInvestigators(data).catch((err) => {
      return next({ status: 500, errors: err });
    });
    if (pi === undefined) {
      return next({ status: 500, errors: "Cannot get authors" });
    }
    if (req.query.acknowledgements === "true") {
      await getAcknowledgements(req, res, data);
      return;
    } else if (req.query.dataAvailability === "true") {
      await getDataAvailability(req, res, data);
      return;
    } else {
      await getCitation(req, res, data, pi);
      return;
    }
  };
}

async function getCitation(req: Request, res: Response, data: RegularFile | ModelFile, pi: Name[]) {
  const datestr = humanReadableDate(data.measurementDate as unknown as string);
  let author = [];
  let publisher = ["ACTRIS Cloud Remote Sensing Unit"];
  if (data instanceof ModelFile) {
    author = [{ first_name: "Ewan", last_name: "O'Connor" }];
  } else {
    author = pi.map((a: Name) => ({ first_name: a.first_name, last_name: a.last_name }));
  }
  const prod = data.product.humanReadableName;
  const site = data.site.humanReadableName;
  const title = `${prod} data from ${site} on ${datestr}`;
  const [year, month, day] = yearMonthDay(data.updatedAt as unknown as string);
  const citekey = `${capitalize(data.site.id)}${capitalize(data.product.id)}${year}${month}${day}`;
  const url = data.pid ? data.pid : getCloudnetUrl(data.uuid);
  const note = data.volatile ? "Data is volatile and may be updated in the future" : "";
  const citation = {
    citekey: citekey,
    author: author,
    publisher: publisher,
    title: title,
    year: year,
    month: month,
    day: day,
    url: url,
    urldate: todayIsoString(),
    note: note,
  };
  if (req.query.format && req.query.format === "bibtex") {
    res.setHeader("Content-type", "text/plain");
    //res.setHeader("Content-type", "application/x-bibtex");
    res.setHeader("Content-Disposition", `inline; filename="${citekey}.bib"`);
    const bibtexstr = citation2bibtex(citation);
    res.send(bibtexstr);
    return;
  } else if (req.query.format && req.query.format === "ris") {
    res.setHeader("Content-type", "text/plain");
    //res.setHeader("Content-type", "application/x-research-info-systems");
    res.setHeader("Content-Disposition", `inline; filename="${citekey}.ris"`);
    const risstr = citation2ris(citation);
    res.send(risstr);
    return;
  } else if (req.query.format && req.query.format === "html") {
    res.setHeader("Content-type", "text/plain");
    const htmlstr = citation2html(citation);
    res.send(htmlstr);
  } else {
    res.json(citation);
    return;
  }
}

async function getDataAvailability(req: Request, res: Response, data: RegularFile | ModelFile) {
  let avail = "";
  if (data.volatile) {
    avail = commonDataAvailabilityVolatileHtml(data);
  } else {
    avail = commonDataAvailabilityStableHtml(data);
  }
  res.send(avail);
}

async function getAcknowledgements(req: Request, res: Response, data: RegularFile | ModelFile) {
  let commonAck = commonAcknowledgement();
  let ack = [];
  if (data instanceof ModelFile) {
    let modelAck = data.model.citations.map((r) => r.acknowledgements);
    ack = [commonAck].concat(modelAck);
  } else {
    let siteAck = data.site.citations.map((r) => r.acknowledgements);
    ack = [commonAck].concat(siteAck);
  }
  let ackstr = ack.join(" ");
  if (req.query.format && req.query.format === "html") {
    res.send(urls2HtmlLinks(ackstr));
    return;
  } else if (req.query.format && req.query.format === "plain") {
    res.setHeader("Content-type", "text/plain");
    res.send(ack.join(" "));
    return;
  } else {
    res.json(ack);
    return;
  }
}

async function getPrincipalInvestigators(data: RegularFile | ModelFile): Promise<Name[]> {
  const actrisId = data.site.actrisId;
  let names: Name[] = [];
  if (actrisId) {
    const response = await axios.get(`${LABELLING_URL}/${actrisId}/contacts`, {
      params: {
        date: data.measurementDate,
        role: "pi",
      },
    });
    if (response !== undefined) {
      names = response.data.map((e: any) => ({ first_name: e.first_name, last_name: e.last_name }));
    }
  }
  return names;
}

function capitalize(str: string): string {
  return str.slice(0, 1).toUpperCase() + str.slice(1);
}

function humanReadableDate(isodate: string) {
  const date = new Date(isodate);
  const year = new Intl.DateTimeFormat("en-GB", { year: "numeric" }).format(date);
  const month = new Intl.DateTimeFormat("en-GB", { month: "long" }).format(date);
  const day = new Intl.DateTimeFormat("en-GB", { day: "numeric" }).format(date);
  return `${day} ${month} ${year}`;
}

function yearMonthDay(isodate: string) {
  const date = new Date(isodate);
  const year = new Intl.DateTimeFormat("en-GB", { year: "numeric" }).format(date);
  const month = new Intl.DateTimeFormat("en-GB", { month: "2-digit" }).format(date);
  const day = new Intl.DateTimeFormat("en-GB", { day: "2-digit" }).format(date);
  return [year, month, day];
}

function citation2bibtex(c: Citation) {
  const author = c.author.map((a) => `${a.last_name}, ${a.first_name}`).join(" and ");
  const publisher = c.publisher.join(", ");
  const note = c.note.length > 0 ? `note = {${c.note}},` : "";
  return `\
@misc{${c.citekey},
  author = {${author}},
  title = {${c.title}},
  year = {${c.year}},
  month = {${c.month}},
  day = {${c.day}},
  url = {${c.url}},
  urldate = {${c.urldate}},
  publisher = {${publisher}},
  licence = {${getCopyrightUrl()}},
  ${note}
}`.replace(/^\s*\n/gm, "");
}

function citation2html(c: Citation) {
  let author = c.author.map((a) => `${a.last_name}, ${a.first_name}`).join(" and ");
  if (author.length > 0) {
    author = author.concat(": ");
  }
  const url = c.url.length > 0 ? urls2HtmlLinks(c.url).concat(", ") : "";
  let publisher = c.publisher.join(", ");
  publisher = publisher.length > 0 ? publisher.concat(", ") : "";
  return `${author}${c.title}, ${publisher}${url}${c.year}`;
}

function citation2ris(c: Citation) {
  const endl = "\r\n";
  let author = c.author.map((a) => `AU  - ${a.last_name}, ${a.first_name}`).join("\r\n");
  let publisher = c.publisher.join(", ");
  if (author == "") {
    author = "AU  - ";
  }
  return `\
TY  - DATA${endl}\
${author}${endl}\
T1  - ${c.title}${endl}\
PY  - ${c.year}/${c.month}/${c.day}${endl}\
PB  - ${publisher}${endl}\
UR  - ${c.url}${endl}\
Y2  - ${c.urldate}${endl}\
N1  - ${c.note}${endl}`;
}

function commonAcknowledgement() {
  let ackstr = `
    The Finnish Meteorological Institute for data processing.
  `;
  return ackstr.replace(/\s\s+/g, " ").trim();
}

function commonDataAvailabilityVolatileHtml(data: RegularFile | ModelFile) {
  const url = getCloudnetUrl(data.uuid);
  const datastr = `\
      The data is available at
      the Aerosol, Clouds and Trace Gases Research Infrastructure (ACTRIS) Data Centre :
      <a href="${url}">${url}</a>. The data is volatile and may be updated in the future.`;
  return datastr.replace(/\s\s+/g, " ").trim();
}
function commonDataAvailabilityStableHtml(data: RegularFile | ModelFile) {
  const pid = data.pid;
  const datastr = `\
      The data is available at
      the Aerosol, Clouds and Trace Gases Research Infrastructure (ACTRIS) Data Centre:
      <a href="${pid}">${pid}</a>.`;
  return datastr.replace(/\s\s+/g, " ").trim();
}

function urls2HtmlLinks(str: string) {
  const urlRegex = /((https?:\/\/)[a-z0-9]+([-.]{1}[a-z0-9]+)*\.[a-z]{2,5}(:[0-9]{1,5})?(\/.*)?(?<!\.))/g;
  return str.replace(urlRegex, '<a href="$1">$1</a>');
}

function getCopyrightUrl() {
  return `https://creativecommons.org/licenses/by/4.0/`;
}

function getCloudnetUrl(uuid: string) {
  return `https://cloudnet.fmi.fi/file/${uuid}`;
}

function todayIsoString() {
  const today = new Date();
  return today.toISOString().split("T")[0];
}
