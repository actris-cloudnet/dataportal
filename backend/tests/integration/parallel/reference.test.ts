import { backendPublicUrl } from "../../lib";
import axios from "axios";
import { describe, expect, it } from "@jest/globals";
import { html2txt } from "../../../src/routes/reference";

const url = `${backendPublicUrl}`;
const uuids = [
  // Regular files
  "2bb32746-faf0-4057-9076-ed2e698dcf36",
  "1bb32746-faf0-4057-9076-ed2e698dcf36",
  "b6de8cf4-8825-47b0-aaa9-4fd413bbb0d7",
  "f036da43-c19c-4832-99f9-6cc88f3255c5",
  // Model files
  "a45a2e9a-e39d-4af2-9798-5ea0fadf041e",
  // Legacy files
  "0afca83a-7b6b-4288-82f6-a59685346617",
  // Collections of regular and model files
  "48092c00-161d-4ca2-a29d-628cf8e960f6",
  "7a6c6675-af65-4650-a20c-e061344393d1",
  // Collection of legacy files
  "6e145ec8-7d66-4a5b-9c5e-c1fdfadafc6b",
];

describe("GET /api/reference", () => {
  const examples = [
    {
      input: "This is a link.",
      expected: "This is a link.",
    },
    {
      input: "This is a link to <a href='https://example.com'>a website.</a>",
      expected: "This is a link to a website.",
    },
    {
      input: "This is a link to <a href='https://example.com'>https://example.com</a>.",
      expected: "This is a link to https://example.com.",
    },
  ];

  for (const example of examples) {
    it("converts HTML to text", () => {
      expect(html2txt(example.input)).toBe(example.expected);
    });
  }
});

describe("GET /api/reference", () => {
  for (const uuid of uuids) {
    describe(`/${uuid}`, () => {
      for (const format of ["html", "bibtex", "ris", "txt"]) {
        it(`outputs citation as ${format}`, async () => {
          const res = await axios.get(`${url}reference/${uuid}/citation`, {
            params: { format: format },
          });
          expect(res.data).toMatchSnapshot();
        });
      }

      for (const format of ["html", "txt"]) {
        it(`outputs acknowledgements as ${format}`, async () => {
          const res = await axios.get(`${url}reference/${uuid}/acknowledgements`, {
            params: { format: format },
          });
          expect(res.data).toMatchSnapshot();
        });
      }

      for (const format of ["html", "txt"]) {
        it(`outputs data availability`, async () => {
          const res = await axios.get(`${url}reference/${uuid}/data-availability`, {
            params: { format: format },
          });
          expect(res.data).toMatchSnapshot();
        });
      }
    });
  }
});
