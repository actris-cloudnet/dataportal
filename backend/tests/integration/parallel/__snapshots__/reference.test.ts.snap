// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GET /api/reference /0afca83a-7b6b-4288-82f6-a59685346617 outputs acknowledgements 1`] = `"We acknowledge ACTRIS and Finnish Meteorological Institute for providing the data set which is available for download from <a href="http://localhost:8080">http://localhost:8080</a>."`;

exports[`GET /api/reference /0afca83a-7b6b-4288-82f6-a59685346617 outputs citation as bibtex 1`] = `
"@misc{clu2020,
  author = {CLU},
  title = {Classification data from New York on 16 July 2009},
  year = {2020},
  url = {http://localhost:8080/file/0afca83a-7b6b-4288-82f6-a59685346617},
  publisher = {ACTRIS Cloud remote sensing data centre unit (CLU)},
  licence = {https://creativecommons.org/licenses/by/4.0/}
}"
`;

exports[`GET /api/reference /0afca83a-7b6b-4288-82f6-a59685346617 outputs citation as html 1`] = `"CLU (2020). Classification data from New York on 16 July 2009. ACTRIS Cloud remote sensing data centre unit (CLU). <a href="http://localhost:8080/file/0afca83a-7b6b-4288-82f6-a59685346617">http://localhost:8080/file/0afca83a-7b6b-4288-82f6-a59685346617</a>"`;

exports[`GET /api/reference /0afca83a-7b6b-4288-82f6-a59685346617 outputs citation as ris 1`] = `
"TY  - DATA
AU  - CLU
T1  - Classification data from New York on 16 July 2009
PY  - 2020
PB  - ACTRIS Cloud remote sensing data centre unit (CLU)
UR  - http://localhost:8080/file/0afca83a-7b6b-4288-82f6-a59685346617
N1  - 
ER  -"
`;

exports[`GET /api/reference /0afca83a-7b6b-4288-82f6-a59685346617 outputs data availability 1`] = `"The data used in this study are generated by the Aerosol, Clouds and Trace Gases Research Infrastructure (ACTRIS) and are available from the ACTRIS Data Centre using the following link: <a href="http://localhost:8080/file/0afca83a-7b6b-4288-82f6-a59685346617">http://localhost:8080/file/0afca83a-7b6b-4288-82f6-a59685346617</a>."`;

exports[`GET /api/reference /1bb32746-faf0-4057-9076-ed2e698dcf36 outputs acknowledgements 1`] = `"We acknowledge ACTRIS and Finnish Meteorological Institute for providing the data set which is available for download from <a href="http://localhost:8080">http://localhost:8080</a>. Bucharest test citation. We acknowledge ECMWF citation text."`;

exports[`GET /api/reference /1bb32746-faf0-4057-9076-ed2e698dcf36 outputs citation as bibtex 1`] = `
"@misc{kirk2020,
  author = {Kirk, James Tiberius and Picard, Jean-Luc and O'Connor, Ewan and Buchareström, Björn and Bucharinen, Benjamin},
  title = {Categorize data from Bucharest on 16 July 2019},
  year = {2020},
  url = {http://localhost:8080/file/1bb32746-faf0-4057-9076-ed2e698dcf36},
  publisher = {ACTRIS Cloud remote sensing data centre unit (CLU)},
  licence = {https://creativecommons.org/licenses/by/4.0/}
}"
`;

exports[`GET /api/reference /1bb32746-faf0-4057-9076-ed2e698dcf36 outputs citation as html 1`] = `"Kirk, J. T., Picard, J.-L., O'Connor, E., Buchareström, B., & Bucharinen, B. (2020). Categorize data from Bucharest on 16 July 2019. ACTRIS Cloud remote sensing data centre unit (CLU). <a href="http://localhost:8080/file/1bb32746-faf0-4057-9076-ed2e698dcf36">http://localhost:8080/file/1bb32746-faf0-4057-9076-ed2e698dcf36</a>"`;

exports[`GET /api/reference /1bb32746-faf0-4057-9076-ed2e698dcf36 outputs citation as ris 1`] = `
"TY  - DATA
AU  - Kirk, James Tiberius
AU  - Picard, Jean-Luc
AU  - O'Connor, Ewan
AU  - Buchareström, Björn
AU  - Bucharinen, Benjamin
T1  - Categorize data from Bucharest on 16 July 2019
PY  - 2020
PB  - ACTRIS Cloud remote sensing data centre unit (CLU)
UR  - http://localhost:8080/file/1bb32746-faf0-4057-9076-ed2e698dcf36
N1  - 
ER  -"
`;

exports[`GET /api/reference /1bb32746-faf0-4057-9076-ed2e698dcf36 outputs data availability 1`] = `"The data used in this study are generated by the Aerosol, Clouds and Trace Gases Research Infrastructure (ACTRIS) and are available from the ACTRIS Data Centre using the following link: <a href="http://localhost:8080/file/1bb32746-faf0-4057-9076-ed2e698dcf36">http://localhost:8080/file/1bb32746-faf0-4057-9076-ed2e698dcf36</a>."`;

exports[`GET /api/reference /2bb32746-faf0-4057-9076-ed2e698dcf36 outputs acknowledgements 1`] = `"We acknowledge ACTRIS and Finnish Meteorological Institute for providing the data set which is available for download from <a href="http://localhost:8080">http://localhost:8080</a>. Bucharest test citation. We acknowledge ECMWF citation text."`;

exports[`GET /api/reference /2bb32746-faf0-4057-9076-ed2e698dcf36 outputs citation as bibtex 1`] = `
"@misc{kirk2020,
  author = {Kirk, James Tiberius and Picard, Jean-Luc and O'Connor, Ewan and Buchareström, Björn and Bucharinen, Benjamin},
  title = {Classification data from Bucharest on 16 July 2019},
  year = {2020},
  url = {http://localhost:8080/file/2bb32746-faf0-4057-9076-ed2e698dcf36},
  publisher = {ACTRIS Cloud remote sensing data centre unit (CLU)},
  licence = {https://creativecommons.org/licenses/by/4.0/}
}"
`;

exports[`GET /api/reference /2bb32746-faf0-4057-9076-ed2e698dcf36 outputs citation as html 1`] = `"Kirk, J. T., Picard, J.-L., O'Connor, E., Buchareström, B., & Bucharinen, B. (2020). Classification data from Bucharest on 16 July 2019. ACTRIS Cloud remote sensing data centre unit (CLU). <a href="http://localhost:8080/file/2bb32746-faf0-4057-9076-ed2e698dcf36">http://localhost:8080/file/2bb32746-faf0-4057-9076-ed2e698dcf36</a>"`;

exports[`GET /api/reference /2bb32746-faf0-4057-9076-ed2e698dcf36 outputs citation as ris 1`] = `
"TY  - DATA
AU  - Kirk, James Tiberius
AU  - Picard, Jean-Luc
AU  - O'Connor, Ewan
AU  - Buchareström, Björn
AU  - Bucharinen, Benjamin
T1  - Classification data from Bucharest on 16 July 2019
PY  - 2020
PB  - ACTRIS Cloud remote sensing data centre unit (CLU)
UR  - http://localhost:8080/file/2bb32746-faf0-4057-9076-ed2e698dcf36
N1  - 
ER  -"
`;

exports[`GET /api/reference /2bb32746-faf0-4057-9076-ed2e698dcf36 outputs data availability 1`] = `"The data used in this study are generated by the Aerosol, Clouds and Trace Gases Research Infrastructure (ACTRIS) and are available from the ACTRIS Data Centre using the following link: <a href="http://localhost:8080/file/2bb32746-faf0-4057-9076-ed2e698dcf36">http://localhost:8080/file/2bb32746-faf0-4057-9076-ed2e698dcf36</a>."`;

exports[`GET /api/reference /6e145ec8-7d66-4a5b-9c5e-c1fdfadafc6b outputs acknowledgements 1`] = `"We acknowledge ACTRIS and Finnish Meteorological Institute for providing the data set which is available for download from <a href="http://localhost:8080">http://localhost:8080</a>."`;

exports[`GET /api/reference /6e145ec8-7d66-4a5b-9c5e-c1fdfadafc6b outputs citation as bibtex 1`] = `
"@misc{clu2020,
  author = {CLU},
  title = {Custom collection of classification data from New York on 16 Jul 2009},
  year = {2020},
  url = {http://localhost:8080/collection/6e145ec8-7d66-4a5b-9c5e-c1fdfadafc6b},
  publisher = {ACTRIS Cloud remote sensing data centre unit (CLU)},
  licence = {https://creativecommons.org/licenses/by/4.0/}
}"
`;

exports[`GET /api/reference /6e145ec8-7d66-4a5b-9c5e-c1fdfadafc6b outputs citation as html 1`] = `"CLU (2020). Custom collection of classification data from New York on 16 Jul 2009. ACTRIS Cloud remote sensing data centre unit (CLU). <a href="http://localhost:8080/collection/6e145ec8-7d66-4a5b-9c5e-c1fdfadafc6b">http://localhost:8080/collection/6e145ec8-7d66-4a5b-9c5e-c1fdfadafc6b</a>"`;

exports[`GET /api/reference /6e145ec8-7d66-4a5b-9c5e-c1fdfadafc6b outputs citation as ris 1`] = `
"TY  - DATA
AU  - CLU
T1  - Custom collection of classification data from New York on 16 Jul 2009
PY  - 2020
PB  - ACTRIS Cloud remote sensing data centre unit (CLU)
UR  - http://localhost:8080/collection/6e145ec8-7d66-4a5b-9c5e-c1fdfadafc6b
N1  - 
ER  -"
`;

exports[`GET /api/reference /6e145ec8-7d66-4a5b-9c5e-c1fdfadafc6b outputs data availability 1`] = `"The data used in this study are generated by the Aerosol, Clouds and Trace Gases Research Infrastructure (ACTRIS) and are available from the ACTRIS Data Centre using the following link: <a href="http://localhost:8080/collection/6e145ec8-7d66-4a5b-9c5e-c1fdfadafc6b">http://localhost:8080/collection/6e145ec8-7d66-4a5b-9c5e-c1fdfadafc6b</a>."`;

exports[`GET /api/reference /7a6c6675-af65-4650-a20c-e061344393d1 outputs acknowledgements 1`] = `"We acknowledge ACTRIS and Finnish Meteorological Institute for providing the data set which is available for download from <a href="http://localhost:8080">http://localhost:8080</a>. Bucharest test citation. We acknowledge ECMWF citation text."`;

exports[`GET /api/reference /7a6c6675-af65-4650-a20c-e061344393d1 outputs citation as bibtex 1`] = `
"@misc{kirk2020,
  author = {Kirk, James Tiberius and Picard, Jean-Luc and O'Connor, Ewan and Buchareström, Björn and Bucharinen, Benjamin},
  title = {Custom collection of classification data from Bucharest on 16 Jul 2019},
  year = {2020},
  url = {http://localhost:8080/collection/7a6c6675-af65-4650-a20c-e061344393d1},
  publisher = {ACTRIS Cloud remote sensing data centre unit (CLU)},
  licence = {https://creativecommons.org/licenses/by/4.0/}
}"
`;

exports[`GET /api/reference /7a6c6675-af65-4650-a20c-e061344393d1 outputs citation as html 1`] = `"Kirk, J. T., Picard, J.-L., O'Connor, E., Buchareström, B., & Bucharinen, B. (2020). Custom collection of classification data from Bucharest on 16 Jul 2019. ACTRIS Cloud remote sensing data centre unit (CLU). <a href="http://localhost:8080/collection/7a6c6675-af65-4650-a20c-e061344393d1">http://localhost:8080/collection/7a6c6675-af65-4650-a20c-e061344393d1</a>"`;

exports[`GET /api/reference /7a6c6675-af65-4650-a20c-e061344393d1 outputs citation as ris 1`] = `
"TY  - DATA
AU  - Kirk, James Tiberius
AU  - Picard, Jean-Luc
AU  - O'Connor, Ewan
AU  - Buchareström, Björn
AU  - Bucharinen, Benjamin
T1  - Custom collection of classification data from Bucharest on 16 Jul 2019
PY  - 2020
PB  - ACTRIS Cloud remote sensing data centre unit (CLU)
UR  - http://localhost:8080/collection/7a6c6675-af65-4650-a20c-e061344393d1
N1  - 
ER  -"
`;

exports[`GET /api/reference /7a6c6675-af65-4650-a20c-e061344393d1 outputs data availability 1`] = `"The data used in this study are generated by the Aerosol, Clouds and Trace Gases Research Infrastructure (ACTRIS) and are available from the ACTRIS Data Centre using the following link: <a href="http://localhost:8080/collection/7a6c6675-af65-4650-a20c-e061344393d1">http://localhost:8080/collection/7a6c6675-af65-4650-a20c-e061344393d1</a>."`;

exports[`GET /api/reference /48092c00-161d-4ca2-a29d-628cf8e960f6 outputs acknowledgements 1`] = `"We acknowledge ACTRIS and Finnish Meteorological Institute for providing the data set which is available for download from <a href="http://localhost:8080">http://localhost:8080</a>. Hyytiälä test citation."`;

exports[`GET /api/reference /48092c00-161d-4ca2-a29d-628cf8e960f6 outputs citation as bibtex 1`] = `
"@misc{clu2020,
  author = {CLU},
  title = {Custom collection of model, and radar data from Hyytiälä, and Mace Head between 5 Dec 2014 and 1 Sep 2019},
  year = {2020},
  url = {http://localhost:8080/collection/48092c00-161d-4ca2-a29d-628cf8e960f6},
  publisher = {ACTRIS Cloud remote sensing data centre unit (CLU)},
  licence = {https://creativecommons.org/licenses/by/4.0/}
}"
`;

exports[`GET /api/reference /48092c00-161d-4ca2-a29d-628cf8e960f6 outputs citation as html 1`] = `"CLU (2020). Custom collection of model, and radar data from Hyytiälä, and Mace Head between 5 Dec 2014 and 1 Sep 2019. ACTRIS Cloud remote sensing data centre unit (CLU). <a href="http://localhost:8080/collection/48092c00-161d-4ca2-a29d-628cf8e960f6">http://localhost:8080/collection/48092c00-161d-4ca2-a29d-628cf8e960f6</a>"`;

exports[`GET /api/reference /48092c00-161d-4ca2-a29d-628cf8e960f6 outputs citation as ris 1`] = `
"TY  - DATA
AU  - CLU
T1  - Custom collection of model, and radar data from Hyytiälä, and Mace Head between 5 Dec 2014 and 1 Sep 2019
PY  - 2020
PB  - ACTRIS Cloud remote sensing data centre unit (CLU)
UR  - http://localhost:8080/collection/48092c00-161d-4ca2-a29d-628cf8e960f6
N1  - 
ER  -"
`;

exports[`GET /api/reference /48092c00-161d-4ca2-a29d-628cf8e960f6 outputs data availability 1`] = `"The data used in this study are generated by the Aerosol, Clouds and Trace Gases Research Infrastructure (ACTRIS) and are available from the ACTRIS Data Centre using the following link: <a href="http://localhost:8080/collection/48092c00-161d-4ca2-a29d-628cf8e960f6">http://localhost:8080/collection/48092c00-161d-4ca2-a29d-628cf8e960f6</a>."`;

exports[`GET /api/reference /a45a2e9a-e39d-4af2-9798-5ea0fadf041e outputs acknowledgements 1`] = `"We acknowledge ACTRIS and Finnish Meteorological Institute for providing the data set which is available for download from <a href="http://localhost:8080">http://localhost:8080</a>. We acknowledge ECMWF citation text."`;

exports[`GET /api/reference /a45a2e9a-e39d-4af2-9798-5ea0fadf041e outputs citation as bibtex 1`] = `
"@misc{o'connor2020,
  author = {O'Connor, Ewan},
  title = {Model data from Bucharest on 16 July 2019},
  year = {2020},
  url = {www.pid.com},
  publisher = {ACTRIS Cloud remote sensing data centre unit (CLU)},
  licence = {https://creativecommons.org/licenses/by/4.0/}
}"
`;

exports[`GET /api/reference /a45a2e9a-e39d-4af2-9798-5ea0fadf041e outputs citation as html 1`] = `"O'Connor, E. (2020). Model data from Bucharest on 16 July 2019. ACTRIS Cloud remote sensing data centre unit (CLU). <a href="www.pid.com">www.pid.com</a>"`;

exports[`GET /api/reference /a45a2e9a-e39d-4af2-9798-5ea0fadf041e outputs citation as ris 1`] = `
"TY  - DATA
AU  - O'Connor, Ewan
T1  - Model data from Bucharest on 16 July 2019
PY  - 2020
PB  - ACTRIS Cloud remote sensing data centre unit (CLU)
UR  - www.pid.com
N1  - 
ER  -"
`;

exports[`GET /api/reference /a45a2e9a-e39d-4af2-9798-5ea0fadf041e outputs data availability 1`] = `"The data used in this study are generated by the Aerosol, Clouds and Trace Gases Research Infrastructure (ACTRIS) and are available from the ACTRIS Data Centre using the following link: <a href="www.pid.com">www.pid.com</a>."`;

exports[`GET /api/reference /b6de8cf4-8825-47b0-aaa9-4fd413bbb0d7 outputs acknowledgements 1`] = `"We acknowledge ACTRIS and Finnish Meteorological Institute for providing the data set which is available for download from <a href="http://localhost:8080">http://localhost:8080</a>. Bucharest test citation."`;

exports[`GET /api/reference /b6de8cf4-8825-47b0-aaa9-4fd413bbb0d7 outputs citation as bibtex 1`] = `
"@misc{picard2020,
  author = {Picard, Jean-Luc and Buchareström, Björn and Bucharinen, Benjamin},
  title = {Lidar data from Bucharest on 16 July 2019},
  year = {2020},
  url = {http://localhost:8080/file/b6de8cf4-8825-47b0-aaa9-4fd413bbb0d7},
  publisher = {ACTRIS Cloud remote sensing data centre unit (CLU)},
  licence = {https://creativecommons.org/licenses/by/4.0/}
}"
`;

exports[`GET /api/reference /b6de8cf4-8825-47b0-aaa9-4fd413bbb0d7 outputs citation as html 1`] = `"Picard, J.-L., Buchareström, B., & Bucharinen, B. (2020). Lidar data from Bucharest on 16 July 2019. ACTRIS Cloud remote sensing data centre unit (CLU). <a href="http://localhost:8080/file/b6de8cf4-8825-47b0-aaa9-4fd413bbb0d7">http://localhost:8080/file/b6de8cf4-8825-47b0-aaa9-4fd413bbb0d7</a>"`;

exports[`GET /api/reference /b6de8cf4-8825-47b0-aaa9-4fd413bbb0d7 outputs citation as ris 1`] = `
"TY  - DATA
AU  - Picard, Jean-Luc
AU  - Buchareström, Björn
AU  - Bucharinen, Benjamin
T1  - Lidar data from Bucharest on 16 July 2019
PY  - 2020
PB  - ACTRIS Cloud remote sensing data centre unit (CLU)
UR  - http://localhost:8080/file/b6de8cf4-8825-47b0-aaa9-4fd413bbb0d7
N1  - 
ER  -"
`;

exports[`GET /api/reference /b6de8cf4-8825-47b0-aaa9-4fd413bbb0d7 outputs data availability 1`] = `"The data used in this study are generated by the Aerosol, Clouds and Trace Gases Research Infrastructure (ACTRIS) and are available from the ACTRIS Data Centre using the following link: <a href="http://localhost:8080/file/b6de8cf4-8825-47b0-aaa9-4fd413bbb0d7">http://localhost:8080/file/b6de8cf4-8825-47b0-aaa9-4fd413bbb0d7</a>."`;

exports[`GET /api/reference /f036da43-c19c-4832-99f9-6cc88f3255c5 outputs acknowledgements 1`] = `"We acknowledge ACTRIS and Finnish Meteorological Institute for providing the data set which is available for download from <a href="http://localhost:8080">http://localhost:8080</a>. Bucharest test citation."`;

exports[`GET /api/reference /f036da43-c19c-4832-99f9-6cc88f3255c5 outputs citation as bibtex 1`] = `
"@misc{kirk2020,
  author = {Kirk, James Tiberius and Buchareström, Björn and Bucharinen, Benjamin},
  title = {Radar data from Bucharest on 16 July 2019},
  year = {2020},
  url = {http://localhost:8080/file/f036da43-c19c-4832-99f9-6cc88f3255c5},
  publisher = {ACTRIS Cloud remote sensing data centre unit (CLU)},
  licence = {https://creativecommons.org/licenses/by/4.0/}
}"
`;

exports[`GET /api/reference /f036da43-c19c-4832-99f9-6cc88f3255c5 outputs citation as html 1`] = `"Kirk, J. T., Buchareström, B., & Bucharinen, B. (2020). Radar data from Bucharest on 16 July 2019. ACTRIS Cloud remote sensing data centre unit (CLU). <a href="http://localhost:8080/file/f036da43-c19c-4832-99f9-6cc88f3255c5">http://localhost:8080/file/f036da43-c19c-4832-99f9-6cc88f3255c5</a>"`;

exports[`GET /api/reference /f036da43-c19c-4832-99f9-6cc88f3255c5 outputs citation as ris 1`] = `
"TY  - DATA
AU  - Kirk, James Tiberius
AU  - Buchareström, Björn
AU  - Bucharinen, Benjamin
T1  - Radar data from Bucharest on 16 July 2019
PY  - 2020
PB  - ACTRIS Cloud remote sensing data centre unit (CLU)
UR  - http://localhost:8080/file/f036da43-c19c-4832-99f9-6cc88f3255c5
N1  - 
ER  -"
`;

exports[`GET /api/reference /f036da43-c19c-4832-99f9-6cc88f3255c5 outputs data availability 1`] = `"The data used in this study are generated by the Aerosol, Clouds and Trace Gases Research Infrastructure (ACTRIS) and are available from the ACTRIS Data Centre using the following link: <a href="http://localhost:8080/file/f036da43-c19c-4832-99f9-6cc88f3255c5">http://localhost:8080/file/f036da43-c19c-4832-99f9-6cc88f3255c5</a>."`;
